modules/crm_tools/crm_tools.module

```diff
@@ -0,0 +1,344 @@
+  /** @var \Drupal\Core\Entity\EntityTypeInterface $role_type */
+  $role_type = $entity_types['user_role'];
+  $role_type->setHandlerClass('storage', 'Drupal\crm_tools\AdvancedRoleStorage');
+}
+
+/**
+ * Prepares variables for a crm hat icon.
+ *
+ * Default template: crm-tools-hat.html.twig.
+ *
+ * @param array $variables
+ *   An associative array containing:
+ *   - attributes: HTML markup attributes for the content wrapper.
+ *   - role: The machine name for the role to be rendered.
+ */
+function template_preprocess_crm_tools_hat(array &$variables) {
+  /* @var \Drupal\user\Entity\Role $role */
+  $role = \Drupal::entityTypeManager()->getStorage('user_role')->load($variables['role']);
+  $variables['label'] = $role->label();
+  $variables['color'] = '#' . $role->getThirdPartySetting('crm_tools', 'crm_tools_color') ?: 'blue';
+  $variables['icon'] = 'fa-' . $role->getThirdPartySetting('crm_tools', 'crm_tools_icon') ?: 'fa-user';
+}
+
+/**
+ * Implements hook_form_BASE_FORM_ID_alter() for user_form.
+ */
+function crm_tools_form_user_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if (isset($form['account']['roles'])) {
+    $form['account']['roles']['#attached']['library'][] = 'crm_tools/admin-forms';
+    $options = [];
+    /* @var \Drupal\crm_tools\AdvancedRoleStorageInterface $role_storage */
+    $role_storage = \Drupal::entityTypeManager()->getStorage('user_role');
+    if ($roles = $role_storage->loadTree()) {
+      foreach ($roles as $role) {
+        if ($role->id() == RoleInterface::ANONYMOUS_ID) {
+          continue;
+        }
+        $options[$role->id()] = str_repeat('-', $role->get('depth')) . Html::escape($role->label());
+      }
+      $form['account']['roles']['#attributes']['class'][] = 'crm_tools_roles';
+      $form['account']['roles']['#options'] = $options;
+      $form['account']['roles']['#process'] = \Drupal::service('element_info')->getInfoProperty('checkboxes', '#process', []);
+      $form['account']['roles']['#process'][] = 'crm_tools_user_roles_checkboxes_process';
+    }
+  }
+}
+
+/**
+ * Processes a checkboxes form element.
+ *
+ * @see crm_tools_form_user_form_alter()
+ */
+function crm_tools_user_roles_checkboxes_process(&$element, FormStateInterface $form_state, &$complete_form) {
+  /* @var \Drupal\crm_tools\AdvancedRoleStorageInterface $role_storage */
+  $role_storage = \Drupal::entityTypeManager()->getStorage('user_role');
+  if ($roles = $role_storage->loadTree(0, NULL, TRUE)) {
+    foreach (Element::children($element) as $id) {
+      if (!empty($element[$id]) && !empty($roles[$id])) {
+        $parent = is_array($roles[$id]->get('parents')) ? $roles[$id]->get('parents')[0] : $roles[$id]->get('parents');
+        $element[$id]['#attributes']['data-crm-tools-parent'] = $parent;
+      }
+
+      $children = implode(':', array_keys($role_storage->loadChildren($id)));
+      $element[$id]['#attributes']['data-crm-tools-children'] = $children;
```

Comment:
Is there any reason to do this in a process rather than inline in the form alter, as we've already loaded the roles etc...

Reply:
Yes, the elements that we are adding attributes to are build by the checkboxes render element in Checkboxes::processCheckboxes().